### DISCUSSION TITLE
.gitignore and .eleventyignore don’t support negative patterns

### DISCUSSION THREAD

**Discussion by veleek** [11ty/eleventy#4026] [Enhancement Queue]
**Describe the bug**
`.eleventyignore` files that use negative ignore patterns return unexpected results because eleventy modifies the paths a bit.

**To Reproduce**
1. Create an eleventy ignore file with a regular pattern and a negative pattern
    ```
    site/**/*
    !site/events/2018/**
    ```
2. Run eleventy ignore
3. Note that no files are copied

**Expected behavior**
Negative patterns should allow you to re-include ignored paths

**Environment:**
 - OS and Version: Windows
 - Eleventy Version: 0.9.0

**Additional context**
Eleventy modifies the patterns in the ignore file by prefixing a relative path header (`./`) and then adding the `!` to the beginning.  As a result, the example above (`!site/events/2018/**`) gets turned into `!./!site/events/2018/**` when being passed to fast-glob.  Fast glob processes negative patterns separately so it pulls that pattern out and removes the ! and actually strips the ./ before using it, so it ends up using `!site/events/2018/**` as the negative pattern and this basically says ignore anything that's not a match for `site/events/2018`.  Since `site/events` is not a match for this, the whole folder gets excluded and fast-glob never scans any deeper for files that SHOULD be included.

I haven't had a chance to try it out yet, but this problem could potentially be solve by not modifying the patterns in `.eleventyignore` and using the new `ignore` parameter to fast-glob which handles the negation for you.  I haven't looked too closely at what modifications eleventy does to see if they're necessary for some of the output processing so hopefully it's just not necessary.

Note: This is a new issue related to the upgrade to fast-glob 3.X.  Definitely prefer to use the new version, but would love to have it continue to use .eleventyignore the same way I was before.

**Comment by coolsoftwaretyler** [comment]
*Edit*

Maybe not a duplicate, but related to #677?

**Comment by veleek** [comment]
Yeah.  Pretty much a duplicate.  If you don't have an .eleventyignore file it will use the gitignore by default.  But that explains why my searches for existing bugs didn't find anything.

**Comment by zachleat** [comment]
This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!

**Comment by zachleat** [comment]
Placing this into the enhancement queue. #677 is a related issue to this but not quite the same.

As info, v0.10.0 will include a debug warning if it encounters one of these entries.

**Comment by alberto** [comment]
I was hit by this. My use case is I use Sass so I want to `.gitignore` the CSS files output directory. But I still want eleventy to process them. I tried using a negative pattern in `.eleventyignore` to include them but it didn't work.

In case it helps, `eslint` does something similar with `.gitignore` and `.eslintignore` and it uses https://github.com/kaelzhang/node-ignore to achieve this behaviour.

For others landing here, I worked around it by setting `eleventyConfig.setUseGitIgnore(false);` and duplicating my ignored files in `.eleventyignore`, except the CSS directory, obviously.

**Comment by zachleat** [comment]
A note for future me:

Two dependencies require some level of support here: `chokidar` and `fast-glob`

In 1.0 we’ll use fast-glob’s `ignore` option instead of negated globs. We were already using chokidar’s `ignored` option here.

These will both need to be tested to know if negated patterns are supported in the `ignore`/`ignored` options.

**Comment by zachleat** [comment]
Related #3680, #3442

**Comment by zachleat** [comment]
Cross link to https://www.npmjs.com/package/parse-gitignore or https://www.npmjs.com/package/@humanwhocodes/gitignore-to-minimatch or https://www.npmjs.com/package/ignore
